### ISSUE TITLE
freeze when running ns2 script, no debug output even before first function call

### ISSUE THREAD

**Issue by klepu** [danielyxie/bitburner#3049]
ahoy!

playing bitnode 1.1 to 1.3 went great, in BN5.1 I encountered this *very* strange bug...

steps to reproduce: 
* load attached save
* try to "run newSpread.js psrv-0 joesguns" in terminal
* expected: at least line 25 should be returned: ns.tprint("INFO: starting up");
* bug: game freezes...

Bitburner v1.4.0 a3c599e6, chromium 90.0.4430.212 (Developer Build) built on Debian 10.9, running on Debian 10.11 (64-bit)

edit: same behavior in firefox 91.6.0esr (64-Bit)
edit2: bitburner version is "a 3 c 5 9 9 e 6" ... don't know why the "6" at the end is omitted

sorry for uploading a file with .txt suffix, github won't accept .json ;)

[bitburnerSave_1646112055_BN5x0.json.txt](https://github.com/danielyxie/bitburner/files/8158991/bitburnerSave_1646112055_BN5x0.json.txt)

**Comment by ChrissiQ** [comment]
Hi @klepu ,

I took a look at your script, and found that the described behaviour is caused by a problem in your code.  The code as written is expected to crash.

You have two issues:
1. `printLogsToTerm()` should not be declared `async`
2.  You have three functions - `hackMiningTarget()`, `weakenMiningTarget()`, and `growMiningTarget()` - which resolve synchronously rather than asynchronously.  This is expected to crash.  `async` functions must actually be asynchronous and always call an `await`.  You have an `await` line in each function, but it won't always be called because it is inside an `if` statement.  You can fix this by calling `await ns.sleep(1)` before returning a function result.

I was able to make your code run with those changes and got results printed to the terminal:

```
newSpread.js: INFO: *** starting weaken #1 ***
newSpread.js: INFO: *** starting grow ***
newSpread.js: INFO: *** starting weaken #2 ***
newSpread.js: INFO: *** starting hack ***
newSpread.js: INFO: *** starting weaken #1 ***
newSpread.js: INFO: *** starting grow ***
newSpread.js: INFO: *** starting weaken #2 ***
newSpread.js: INFO: *** starting hack ***
newSpread.js: INFO: *** starting weaken #1 ***
```

etc.

tl;dr: This is not a bug in the game but rather your code.  Every asynchronous function must actually call `await` when run, not just contain an `await` somewhere.  If it isn't called, the function will crash the game as documented.

**Comment by klepu** [comment]
Okay, that helps quite a lot, thank you!

But still I'd like to understand why line 26 (and stuff up to the first wrong await) doesn't get executed... I'm playing the game to learn JS; until I changed some lines that script worked just fine (besides telling me that some awaits wouldn't be needed, but the script crashed without them). When the game froze I tried to pinpoint the culprit by throwing in lots of tprints to debug, but the immediate freeze (with absolutely *no* output) still baffles me. Any hint to why that (doesn't) happen?

**Comment by ChrissiQ** [comment]
I would guess that nothing runs because of [hosting](https://developer.mozilla.org/en-US/docs/Glossary/Hoisting).  The broken functions get hoisted to the top of the scope, but cause a crash because they aren't written correctly.

I understand it can be hard to know what went wrong if you don't get any error messages and all you get is a crash.  It might make it easier if you assume that a crash from a script in this game is probably because of either an infinite loop or a missing/stray `async/await` declaration.  Unfortunately this has little to do with javascript and is specific to NS so I'm afraid this lesson won't extend outside the game.  NS's treatment of asynchronous code is different than real javascript.  In most other respects it's identical though.  This is just one of those quirks.

**Comment by klepu** [comment]
Darn you're quick! I'll read the links you provided and close this bug. Thanks again! <3
